CesiumUtility/include/CesiumUtility/Result.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <CesiumUtility/ErrorList.h>
+#include <CesiumUtility/IntrusivePointer.h>
+
+#include <optional>
+
+namespace CesiumUtility {
+
+/**
+ * @brief Holds the result of an operation. If the operation succeeds, it will
+ * provide a value. It may also provide errors and warnings.
+ *
+ * @tparam T The type of value included in the result.
+ */
+template <typename T> struct Result {
+  Result(T value_) noexcept : value(std::move(value_)), errors() {}
+
+  Result(T value_, ErrorList errors_) noexcept
+      : value(std::move(value_)), errors(std::move(errors_)) {}
+
+  Result(ErrorList errors_) noexcept : value(), errors(std::move(errors_)) {}
+
+  /**
+   * @brief The value, if the operation succeeded to the point where it can
+   * provide one.
+   *
+   * If a value is not provided because the operation failed, then there should
+   * be at least one error in {@link errors} indicating what went wrong.
+   */
+  std::optional<T> value;
+
+  /**
+   * @brief The errors and warnings that occurred during the operation.
+   *
+   * If a {@link value} is provided, there should not be any errors in this
+   * list, but there may be warnings. If a value is not provided, there should
+   * be at least one error in this list.
+   */
+  ErrorList errors;
+};
```

Comment:
If we indeed end up using this type for more than just this feature, it would be lovely to add a lot of the convenience features you find with Rust's Result type. For example, a `match` method that runs one lambda if Ok, or another lambda if Error. Or a `map_value` method than can turn a `Result<T, E>` into a `Result<T2, E>` using a provided lambda. We don't need to add these now, just thinking out loud.

Reply:
Would it be possible to build on `std::expected`? I think we're already using expected-lite here in native.

Reply:
I actually used `nonstd::expected` in an early iteration of this, but the problem is it doesn't offer a good way to represent "success, with non-fatal warnings." Which is a fairly common phenomenon in cesium-native.

I definitely agree that we would benefit from a little work on ErrorList to improve its ergonomics, though. Starting with a rename, because it's weird to call it `ErrorList` when it also holds warnings.